src/nv/storage/public.rs

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    constants::algorithm::HashingAlgorithm,
+    nv::storage::{NvIndex, NvIndexAttributes},
+    structures::Digest,
+    tss2_esys::{TPM2B_NV_PUBLIC, TPMS_NV_PUBLIC},
+    Error, Result, WrapperErrorKind,
+};
+use log::error;
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Debug, Clone)]
+pub struct NvPublic {
+    index: NvIndex,
+    index_name_alg: HashingAlgorithm,
+    index_attributes: NvIndexAttributes,
+    index_auth_policy: Digest,
+    data_area_size: usize,
+}
+
+impl NvPublic {
+    const STRUCTURE_SIZE: usize = std::mem::size_of::<TPMS_NV_PUBLIC>();
+}
+
+impl TryFrom<TPM2B_NV_PUBLIC> for NvPublic {
+    type Error = Error;
+    fn try_from(tss_nv_public: TPM2B_NV_PUBLIC) -> Result<NvPublic> {
+        // Sanity check
+        if tss_nv_public.size as usize != NvPublic::STRUCTURE_SIZE {
+            error!("Error: Encounted an incorrect size of the TPMS_NV_PUBLIC.");
+            return Err(Error::local_error(WrapperErrorKind::WrongParamSize));
+        }
+        // Parse actual data
+        Ok(NvPublic {
+            index: tss_nv_public.nvPublic.nvIndex.into(),
+            index_name_alg: tss_nv_public.nvPublic.nameAlg.try_into()?,
+            index_attributes: tss_nv_public.nvPublic.attributes.into(),
+            index_auth_policy: tss_nv_public.nvPublic.authPolicy.try_into()?,
+            data_area_size: tss_nv_public.nvPublic.dataSize as usize,
+        })
+    }
+}
+
+impl TryFrom<NvPublic> for TPM2B_NV_PUBLIC {
+    type Error = Error;
+    fn try_from(nv_public: NvPublic) -> Result<TPM2B_NV_PUBLIC> {
+        Ok(TPM2B_NV_PUBLIC {
+            size: NvPublic::STRUCTURE_SIZE as u16,
+            nvPublic: TPMS_NV_PUBLIC {
+                nvIndex: nv_public.index.into(),
+                nameAlg: nv_public.index_name_alg.into(),
+                attributes: nv_public.index_attributes.into(),
+                authPolicy: nv_public.index_auth_policy.try_into()?,
+                dataSize: nv_public.data_area_size as u16,
```

Comment:
Should there be a check if `nv_public.data_area_size` is lesser or equal than `u16::MAX`

Reply:
Yes! It should. Thank you.